WalletWasabi/Services/TransactionFactory.cs

```diff
@@ -0,0 +1,295 @@
+			if (allowedInputs != null) // If allowedInputs are specified then select the coins from them.
+			{
+				if (!allowedInputs.Any())
+				{
+					throw new ArgumentException($"{nameof(allowedInputs)} is not null, but empty.");
+				}
+
+				allowedSmartCoinInputs = AllowUnconfirmed
+					? Coins.Where(x => !x.Unavailable && allowedInputs.Any(y => y.TransactionId == x.TransactionId && y.Index == x.Index)).ToList()
+					: Coins.Where(x => !x.Unavailable && x.Confirmed && allowedInputs.Any(y => y.TransactionId == x.TransactionId && y.Index == x.Index)).ToList();
+
+				// Add those that have the same script, because common ownership is already exposed.
+				// But only if the user didn't click the "max" button. In this case he'd send more money than what he'd think.
+				if (payments.ChangeStrategy != ChangeStrategy.AllRemainingCustom)
+				{
+					var allScripts = allowedSmartCoinInputs.Select(x => x.ScriptPubKey).ToHashSet();
+					foreach (var coin in Coins.Where(x => !x.Unavailable && !allowedSmartCoinInputs.Any(y => x.TransactionId == y.TransactionId && x.Index == y.Index)))
+					{
+						if (!(AllowUnconfirmed || coin.Confirmed))
+						{
+							continue;
+						}
+
+						if (allScripts.Contains(coin.ScriptPubKey))
+						{
+							allowedSmartCoinInputs.Add(coin);
+						}
+					}
+				}
+			}
+			else
+			{
+				allowedSmartCoinInputs = AllowUnconfirmed ? Coins.Where(x => !x.Unavailable).ToList() : Coins.Where(x => !x.Unavailable && x.Confirmed).ToList();
+			}
+
+			// Get and calculate fee
+			Logger.LogInfo("Calculating dynamic transaction fee...");
+
+			TransactionBuilder builder = Network.CreateTransactionBuilder();
+			builder.SetCoinSelector(new SmartCoinSelector(allowedSmartCoinInputs));
+			builder.AddCoins(allowedSmartCoinInputs.Select(c => c.GetCoin()));
+
+			foreach (var request in payments.Requests.Where(x => x.Amount.Type == MoneyRequestType.Value))
+			{
+				var amountRequest = request.Amount;
+
+				builder.Send(request.Destination, amountRequest.Amount);
+				if (amountRequest.SubtractFee)
+				{
+					builder.SubtractFees();
+				}
+			}
+
+			HdPubKey changeHdPubKey = null;
+
+			if (payments.TryGetCustomRequest(out DestinationRequest custChange))
+			{
+				var changeScript = custChange.Destination.ScriptPubKey;
+				changeHdPubKey = KeyManager.GetKeyForScriptPubKey(changeScript);
+
+				var changeStrategy = payments.ChangeStrategy;
+				if (changeStrategy == ChangeStrategy.Custom)
+				{
+					builder.SetChange(changeScript);
+				}
+				else if (changeStrategy == ChangeStrategy.AllRemainingCustom)
+				{
+					builder.SendAllRemaining(changeScript);
+				}
+				else
+				{
+					throw new NotSupportedException(payments.ChangeStrategy.ToString());
+				}
+			}
+			else
+			{
+				KeyManager.AssertCleanKeysIndexed(isInternal: true);
+				KeyManager.AssertLockedInternalKeysIndexed(14);
+				changeHdPubKey = KeyManager.GetKeys(KeyState.Clean, true).RandomElement();
```

Comment:
It seems this line was sometimes taking one of the coins' `hdpubkey` because they were not marked as used.